keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+below to improve the clarity of the example.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --proxy-mode=ipvs
+        - --hostname-override=${NODE_NAME}
+        env:
+          - NODE_NAME
+            valueFrom:
+              fieldRef:
+                fieldPath: spec.nodeName
+```
+
+Below demonstrates how the above may be modified when moving to a shared
+ComponentConfig ConfigMap. Note that the instance-specific parameter is still
+specified via a flag, because the ConfigMap is not appropriate for
+instance-specific parameters.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --config=/etc/config/kube-proxy.yaml
+        - --hostname-override=${NODE_NAME}
+        env:
+        - NODE_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: spec.nodeName
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Finally, when instance-specific configuration is also provided via the
+ComponentConfig approach, using the separate file proposed above.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: instance-config
+        image: something-with-bash
+        command:
+        - bash -c
+        args:
+        - cat << EOF > /etc/config/kube-proxy-instance.yaml
+          kind: InstanceConfiguration
+          apiVersion: kubeproxy.config.k8s.io/v1alpha1
+          hostnameOverride: ${NODE_NAME}
+          EOF
+        env:
+        - NODE_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: spec.nodeName
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --config=/etc/config/kube-proxy.yaml
+        - --instance-config=/etc/config/kube-proxy-instance.yaml
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Although the last example is more verbose, it provides the full benefits of
+the ComponentConfig approach, resulting in a consistent surface and versioning
+policy for the component's parameters.
+
+Sometimes, kube-proxy is configured as a static Pod instead of a DaemonSet.
```

Comment:
remind me: does init containers work in a static pod?

Reply:
I believe so.